### ISSUE TITLE
Unexpected Behavior in Pebble Future Termination: is_alive() Returns True After join() Completes

### ISSUE THREAD

**Issue by Vladyyy** [noxdafox/pebble#140]
Hi,

I have a fork of the Pebble library where I added a method to the Pebble future, allowing it to terminate a process and wait for SIGTERM termination with a timeout. Here's the relevant commit:
https://github.com/noxdafox/pebble/commit/969eb71d6cecdf249f6f2cc51aee96b3679f2767

However, I’m encountering something unexpected that I haven't been able to explain or reproduce with multiprocessing.Process.

I’m running the test program below, which creates a process with a signal handler and then terminates the process using the new method I added:

```
#!/usr/bin/env python3
import time
import signal
import multiprocessing as mp

from pebble import concurrent


def foo():
    def exit_gracefully(signum, frame):
        print("Received signal to terminate. Stopping after current iteration")

    signal.signal(signal.SIGTERM, exit_gracefully)
    time.sleep(4)
    print("Done")


if __name__ == "__main__":
    ctx = mp.get_context('spawn')
    process = concurrent.process(foo, name="foo", context=ctx)

    future = process()
    time.sleep(2)
    future.terminate(100)
    future.result()
```

The issue I’m seeing is that the join() call in the terminate method returns before the 100-second timeout and the is_alive() that's called after returns True.

Here’s the output I’m getting:
```
before join is_alive: foo name: True exitcode: None
Received signal to terminate. Stopping after current iteration
Done
after join is_alive: foo name: True exitcode: None
Terminating process foo (PID: 22279) with timeout 100

```

As you can see after the join call, is_alive returns True and exitcode is None

I know this issue might not be directly related to the Pebble code, as I'm modifying the library, but I would appreciate any insights. I haven't been able to find any references to Python multiprocessing bugs that could explain this behavior.

I'm using Python 3.7.16 and HEAD of the pebble branch.

**Comment by noxdafox** [comment]
Hello,

I am not sure what you are trying to do in here as the `is_alive` method is nowhere to be seen in the above example.
As well, `concurrent.Future` does not have a `terminate` function.

By overwriting `SIGTERM` signal handler, you are preventing the default behaviour from happening. When you issue a `SIGTERM` request, your signal handler will simply print the message you see on screen and do nothing else.

**Comment by Vladyyy** [comment]
>Please use github issues to report bugs you observe in the library and not for guidance on how to program in multiprocessing environments. For these matters, you can use stackoverflow.

Sorry for the inconvenience, the reason I brought this up here is because I suspect it's somehow related to the pebble internals

>By overwriting SIGTERM signal handler, you are preventing the default behaviour from happening. When you issue a SIGTERM request, your signal handler will simply print the message you see on screen and do nothing else.

Agree, but the expectation is that the `join` in the `future.terminate` call should wait for the program to complete. However `join` returns early before the 100 seconds timeout.

**Comment by Vladyyy** [comment]
A slight variation of this behavior is seen here:
https://github.com/noxdafox/pebble/blob/master/pebble/concurrent/process.py#L104

In my experiments, the 'if worker.is_alive():' path gets occasionally taken even for successful completion of worker (since _get_result returns as soon as the worker sends the result but there is nothing waiting for the worker to complete).

**Comment by noxdafox** [comment]
The library does not change the behaviour of `is_alive` or `join` and I seem not to be able to reproduce the above issue.

Please note that `is_alive` can return `True` even when the results from a function have been returned. As the child process might still be alive and not terminated yet due to the asynchronous nature of the execution.

`join` always blocks until the underlying process is alive or a timeout is given. What might interrupt a call to `join` might be a signal to the main process while is blocked waiting on `join`. 

I will add means to make the `SIGTERM` to `SIGKILL` timeout configurable to a certain degree.

**Comment by noxdafox** [comment]
I just pushed a change on the `master` branch which will allow you to adjust how much time to wait before the process is forcefully killed.

```python
import time
import signal

from pebble import CONSTS
from pebble import concurrent

# Wait 10 seconds before forcefully killing the child process
CONSTS.term_timeout = 10

@concurrent.process
def function():
    signal.signal(signal.SIGTERM, signal.SIG_IGN)

    time.sleep(300)

future = function()
print("Cancelling..")
future.cancel()
```

**Comment by noxdafox** [comment]
Issue resolved in `5.1.0`.
